internal/kms/kms.go

```diff
@@ -216,6 +221,344 @@ func (k *Kms) CreateKeys(ctx context.Context, scopeId string, opt ...Option) err
+	// table name, thanks to the scheduler guaranteeing that only one instance of a job
+	// is running at a time, it means we were interrupted in our processing last time
+	// and should simply resume that running run.
+	err := k.reader.LookupWhere(ctx, &run, "is_running=true and table_name!=?", []any{tableName})
+	switch {
+	case err == nil:
+		// Another run was already running
+		return nil
+	case errors.Match(errors.T(errors.RecordNotFound), err):
+		// Great, no running run, lets continue
+	default:
+		return errors.Wrap(ctx, err, op, errors.WithMsg("failed to find any running runs"))
+	}
+	// Find the oldest or currently running run for our tablename
+	run = allocDataKeyVersionDestructionJobRun()
+	rows, err := k.reader.Query(ctx, oldestPendingOrRunningRun, []any{tableName})
+	if err != nil {
+		return errors.Wrap(ctx, err, op, errors.WithMsg("failed to query pending runs for %q", tableName))
+	}
+	defer rows.Close()
+	for rows.Next() {
+		if err := k.reader.ScanRows(ctx, rows, &run); err != nil {
+			return errors.Wrap(ctx, err, op, errors.WithMsg("failed to scan pending run for %q", tableName))
+		}
+	}
+	if run.KeyId == "" {
+		// No queued runs, lets try again later
+		return nil
+	}
+	if !run.IsRunning {
+		// No running job, lets try to become the running job
+		run.IsRunning = true
+		if _, err := k.writer.Update(ctx, &run, []string{"IsRunning"}, nil); err != nil {
+			// Unique error means some other run started running since we last checked.
+			if errors.Match(errors.T(errors.NotUnique), err) {
+				return nil
+			}
+			return errors.Wrap(ctx, err, op, errors.WithMsg("failed to become running run for %q", tableName))
+		}
+	}
+
+	defer func() {
+		if retErr != nil {
+			// Create new context in case we failed because the context was canceled
+			var cancel context.CancelFunc
+			ctx, cancel = context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+		}
+		// Update the progress of this run
+		args := []any{
+			sql.Named("key_id", run.KeyId),
+			sql.Named("table_name", tableName),
+		}
+		if _, err := k.writer.Exec(ctx, fmt.Sprintf(updateCompletedCountQueryTemplate, tableName), args); err != nil {
+			if retErr != nil {
+				// Just emit an error, don't mutate the returned error
+				_ = errors.E(ctx, errors.WithWrap(err))
+			} else {
+				retErr = err
+			}
+		}
+	}()
+
+	rows, err = k.reader.Query(ctx, dataKeyVersionIdScopeIdQuery, []any{run.KeyId})
+	if err != nil {
+		return errors.Wrap(ctx, err, op, errors.WithMsg("failed to get scope id for data key version"))
+	}
+	defer rows.Close()
+	var scopeId string
+	for rows.Next() {
+		if err := k.reader.ScanRows(ctx, rows, &scopeId); err != nil {
+			return errors.Wrap(ctx, err, op, errors.WithMsg("failed to scan scope id for data key version"))
+		}
+	}
+
+	// Call the function to rewrap the data in the table. The progress will be automatically
+	// updated by the deferred function.
+	return rewrapFn(ctx, run.KeyId, scopeId, k.reader, k.writer, k)
```

Comment:
should we wrap errors returned from the rewrapFn?

Reply:
I added a wrapper for this